### PR TITLE
Setup typescript

### PR DESCRIPTION
**Pull request by wadjih-bencheikh18**

### CODE REVIEW

**Comment by wadjih-bencheikh18** [comment]
ml-gsd errors left

**Comment by wadjih-bencheikh18** [comment]
Done

File: package.json [hunk 1]
```diff
@@ -2,24 +2,30 @@
...
   "scripts": {
     "build": "rollup -c && cheminfo-build --root SpectraProcessing",
-    "compile": "rollup -c",
+    "check-types": "tsc --noEmit",
+    "clean": "rimraf lib lib-esm",
     "eslint": "eslint src",
     "eslint-fix": "npm run eslint -- --fix",
     "jscpd": "jscpd -l 10 -i \"**/__tests__/**\" -t 1 src",
-    "prepack": "npm run compile",
+    "prepack": "npm run compile && npm run tsc",
```
Comment (lpatiny) [inline package.json:20]: You should keep only `npm run tsc`. compile script was removed

File: package.json [hunk 2]
```diff
@@ -2,24 +2,30 @@
...
+  "module": "./lib-esm/index.js",
+  "types": "./lib/index.d.ts",
   "files": [
+    "src",
     "lib",
-    "src"
+    "lib-esm"
   ],
   "scripts": {
     "build": "rollup -c && cheminfo-build --root SpectraProcessing",
```
Comment (lpatiny) [inline package.json:14]: This build script allows to generate a browser compatible version.

rollup does not exist in this project anymore. This line should be changed to

`npm run tsc-esm && cheminfo-build --entry lib-esm/index.js --root SpectraProcessing`

File: src/index.ts [hunk 3]
```diff
@@ -129,3 +129,50 @@ export * from './matrix/matrixPQN';
 export * from './matrix/matrixZRescale';
 
 export * from './utils/createSequentialArray';
+
+export type ArrayType = number[] | Float64Array | Float32Array | Uint16Array;
```
Comment (lpatiny) [inline src/index.ts:133]: Please only allow number[] | Float64Array

File: src/index.ts [hunk 4]
```diff
@@ -129,3 +129,50 @@ export * from './matrix/matrixPQN';
...
+  y: ArrayType;
+}
+
+export interface DataX {
+  x: ArrayType;
+  re: ArrayType;
+  im: ArrayType;
+}
+
+export interface Reim {
```
Comment (lpatiny) [inline src/index.ts:158]: DataReIm is define a little bit higher in the file and should be used instead of Reim. The definition of DataReIm should be corrected as well so that both are ArrayType

File: src/index.ts [hunk 5]
```diff
@@ -129,3 +129,50 @@ export * from './matrix/matrixPQN';
...
+  x: number;
+  y: number;
+  index?: number;
+}
+export interface XYPoints {
+  x: ArrayType;
+  y: ArrayType;
+}
+
+export interface DataX {
```
Comment (lpatiny) [inline src/index.ts:152]: Please rename to DataXReIm

File: src/index.ts [hunk 6]
```diff
@@ -129,3 +129,50 @@ export * from './matrix/matrixPQN';
...
+  nbPoints?: number;
+  from: number;
+}
+
+export interface Point {
+  x: number;
+  y: number;
+  index?: number;
+}
+export interface XYPoints {
```
Comment (lpatiny) [inline src/index.ts:147]: Should be renamed to DataXY (that is already defined higher but should be replaced)

**Comment by lpatiny** [comment]
Could you check how to write `typedoc` ? After migration I wonder if we style need to write in the 'jsdoc' the types because typedoc will extract them from the typescript code.

Currently we a using typedoc to build the documentation and you can test it yourself 
`npm i --global typedoc`
`typedoc src/index.ts`

And it seems to me that the types are extracted from the function call (and return) and not from the 'jsdoc'

Please could you dig a little bit in this problem to tell us what is the correct way to do ?

If you check the documentation of xyAlign you will also see that the default value of the options parameters are not displayed which is also problematic

**Comment by lpatiny** [comment]
Concerning xyAlign one possibility is to create a new interface for `options` that can then have @default values. The generated typedoc however does not directly display the default values (you need to click on the Options interface to see the default value which is not that nice).

**Comment by lpatiny** [comment]
A discussion about typedoc: https://blog.cloudflare.com/generating-documentation-for-typescript-projects/

File: src/index.ts (hunk 3 above)
Comment (wadjih-bencheikh18) [reply src/index.ts:133]: but there is unit16Array used in tests

File: package.json (hunk 1 above)
Comment (wadjih-bencheikh18) [reply package.json]: Resolved

File: package.json (hunk 2 above)
Comment (wadjih-bencheikh18) [reply package.json]: Resolved

File: src/index.ts (hunk 3 above)
Comment (wadjih-bencheikh18) [reply src/index.ts:133]: Resolved

File: src/index.ts (hunk 4 above)
Comment (wadjih-bencheikh18) [reply src/index.ts]: Resolved

File: src/index.ts (hunk 5 above)
Comment (wadjih-bencheikh18) [reply src/index.ts]: Resolved

File: src/index.ts (hunk 6 above)
Comment (wadjih-bencheikh18) [reply src/index.ts]: Resolved

File: src/index.ts (hunk 3 above)
Comment (wadjih-bencheikh18) [reply src/index.ts]: Fixed

Comment (wadjih-bencheikh18) [reply src/index.ts]: @lpatiny 
Unit16Array used in:
https://github.com/mljs/spectra-processing/blob/aa228beb14e321cd8d43eddec89c1ad9923d7561/src/x/__tests__/xSum.test.js#L4
Float32Array used in:
https://github.com/mljs/spectra-processing/blob/aa228beb14e321cd8d43eddec89c1ad9923d7561/src/x/__tests__/xSubtract.test.js#L16

**Comment by targos** [comment]
Please delete the `docs` folder. It will be built separately on GitHub.

**Comment by targos** [comment]
You can add it to `.gitignore` if you want.

File: .eslintrc.yml [hunk 7]
```diff
@@ -1,4 +1,4 @@
 plugins: [jsdoc]
-extends: [cheminfo, 'plugin:jsdoc/recommended']
+extends: [cheminfo-typescript, 'plugin:jsdoc/recommended']
```
Comment (targos) [inline .eslintrc.yml:2]:
```suggestion
extends: [cheminfo-typescript, cheminfo-typescript/jsdoc]
```

Our custom config is better for TypeScript support (types should not be duplicated in the jsdoc comment)

File: src/xyArray/__tests__/xyArrayAlign.test.ts [hunk 8]
```diff
@@ -1,4 +1,4 @@
-import { xyArrayAlign } from '../xyArrayAlign.js';
+import { xyArrayAlign } from '../xyArrayAlign';
```
Comment (targos) [inline src/xyArray/__tests__/xyArrayAlign.test.ts:1]: Did you remove the `.js` because of an error? If yes, what was it?

Comment (wadjih-bencheikh18) [reply src/xyArray/__tests__/xyArrayAlign.test.ts:1]: Because the file is no longer .js

File: .eslintrc.yml (hunk 7 above)
Comment (wadjih-bencheikh18) [reply .eslintrc.yml]: fixed

File: src/xyArray/__tests__/xyArrayAlign.test.ts (hunk 8 above)
Comment (wadjih-bencheikh18) [reply src/xyArray/__tests__/xyArrayAlign.test.ts:1]: Unable to resolve path to module '../xyArrayAlign.js'.

File: src/matrix/matrixCenterZMean.ts [hunk 9]
```diff
@@ -1,10 +1,10 @@
 import { ArrayType } from '..';
 
-/**
+/**.
```
Comment (targos) [inline src/matrix/matrixCenterZMean.ts:3]: Did you add this manually? The dot should be at the end of the description, not here.
The `eslint-fix` script (or the vscode-eslint extension if you have set it up to fix on save) should correct it automatically.

Comment (wadjih-bencheikh18) [reply src/matrix/matrixCenterZMean.ts:3]: When i save this point appear

**Comment by wadjih-bencheikh18** [comment]
Node.js CI / nodejs / lint-eslint (pull_request) problem is because we used no typescript package "spline-interpolator"

**Comment by targos** [comment]
> Node.js CI / nodejs / lint-eslint (pull_request) problem is because we used no typescript package "spline-interpolator"

@stropitek I think this is a bug in the new lint config.

File: src/matrix/matrixCenterZMean.ts (hunk 9 above)
Comment (targos) [reply src/matrix/matrixCenterZMean.ts:3]: @lpatiny have you seen this happen before?

**Comment by stropitek** [comment]
> @stropitek I think this is a bug in the new lint config.

just released a patch

File: CHANGELOG.md [hunk 10]
```diff
@@ -2,609 +2,479 @@
```
Comment (targos) [inline CHANGELOG.md:2]: Please revert the changes in this file. It is automatically generated/updated

File: src/index.ts (hunk 3 above)
Comment (targos) [reply src/index.ts:105]: Ping @lpatiny

File: src/matrix/matrixCenterZMean.ts (hunk 9 above)
Comment (targos) [reply src/matrix/matrixCenterZMean.ts:3]: I can reproduce, this is an annoying issue...

Comment (targos) [reply src/matrix/matrixCenterZMean.ts:3]: It is a bug in the ESLint plugin: https://github.com/gajus/eslint-plugin-jsdoc/issues/779
I suggest that you search for `/**.` in the project and manually move the dot to the end of the sentence.

Comment (wadjih-bencheikh18) [reply src/matrix/matrixCenterZMean.ts:3]: okey

Comment (wadjih-bencheikh18) [reply src/matrix/matrixCenterZMean.ts:3]: When i delete it manually => Warning : "Sentence must end with a period."
and when I save the file the point reappears

Comment (targos) [reply src/matrix/matrixCenterZMean.ts:3]: You should not only delete it, but also add a dot at the end of the sentence (i.e move the dot)

Comment (wadjih-bencheikh18) [reply src/matrix/matrixCenterZMean.ts:3]: done

File: CHANGELOG.md (hunk 10 above)
Comment (wadjih-bencheikh18) [reply CHANGELOG.md:2]: done

File: src/matrix/matrixCenterZMean.ts (hunk 9 above)
Comment (wadjih-bencheikh18) [reply src/matrix/matrixCenterZMean.ts]: i got the same warning

**Comment by wadjih-bencheikh18** [comment]
Are there any other issues?

**Comment by targos** [comment]
Waiting for @lpatiny input on https://github.com/mljs/spectra-processing/pull/95#discussion_r739876170

**Comment by lpatiny** [comment]
Please have a look at my previous commit.

We would need when we have only options to use 'inline' jsdoc (when we have just few options) and we should remove the corresponding @param from the jsdoc that becomes redundant.

**Comment by lpatiny** [comment]
For now we should only allow Float64 or Array and use the type definition from: https://github.com/cheminfo/cheminfo-types/blob/main/src/core/DoubleArray.d.ts

DataXY can also be reused from: https://github.com/cheminfo/cheminfo-types/blob/main/src/core/DataXY.d.ts

File: src/index.ts [hunk 11]
```diff
@@ -129,3 +129,50 @@ export * from './matrix/matrixPQN';
...
+}
+
+export interface Info {
+  Q1: number;
+  Q2: number;
+  Q3: number;
+  min: number;
+  max: number;
+}
+export interface Data {
```
[the remark this answers is not in the document]
Comment (wadjih-bencheikh18) [reply src/index.ts]: fixed

File: src/index.ts [hunk 12]
```diff
@@ -129,3 +129,50 @@ export * from './matrix/matrixPQN';
...
+  x: ArrayType;
+  y: ArrayType;
+}
+export interface DataXYZ {
+  x?: ArrayType;
+  y?: ArrayType;
+  z?: ArrayType;
+}
```
[the remark this answers is not in the document]
Comment (wadjih-bencheikh18) [reply src/index.ts:150]: okey

File: src/matrix/matrixPQN.ts [hunk 13]
```diff
@@ -7,39 +7,46 @@ import { Matrix } from 'ml-matrix';
...
  * data: Normalized dataset.
  * medianOfQuotients: The median of quotients of each variables.
  */
-export function matrixPQN(matrix, options = {}) {
+export function matrixPQN(
+  matrix: Float64Array[] | number[][] | Float32Array[],
+  options: { max?: number; min?: number } = {},
+): {
+  data: Float64Array[] | number[][] | Float32Array[];
+  medianOfQuotients: Float64Array | number[] | Float32Array;
```
[the remark this answers is not in the document]
Comment (wadjih-bencheikh18) [reply src/matrix/matrixPQN.ts]: fixed

File: src/matrix/matrixColumnsCorrelation.ts [hunk 14]
```diff
@@ -0,0 +1,28 @@
...
+
+import { xCorrelation } from '../x/xCorrelation';
+/**
+ * Calculates a correlation matrix based on the columns of the initial matrix
+ *
+ * @param {Matrix} [A] - matrix [rows][cols]
+ * @returns {number[][]} result
+ */
+export function matrixColumnsCorrelation(
+  A: Float64Array[] | number[][] | Float32Array[],
```
[the remark this answers is not in the document]
Comment (wadjih-bencheikh18) [reply src/matrix/matrixColumnsCorrelation.ts]: i declared a new type in index.ts

File: src/index.ts [hunk 15]
```diff
@@ -129,3 +129,50 @@ export * from './matrix/matrixPQN';
...
+  re: ArrayType;
+  im: ArrayType;
+}
+
+export interface Reim {
+  re: ArrayType;
+  im: ArrayType;
+}
+
+export interface Info {
```
[the remark this answers is not in the document]
Comment (wadjih-bencheikh18) [reply src/index.ts]: i delete this interface in the last commit

**Comment by wadjih-bencheikh18** [comment]
DataXY & DoubleArray are used in last commits (all types are fixed)
which docs you want me excatly to fix?
